Upgrade to 4.5-RC1 failed due to "Multiple input file properties with name '$1'"

I tried to upgrade Gradle from 4.3 to 4.5-RC1, I got below error when I run below command.
seems this error related to https://github.com/gradle/gradle/pull/3720, how can I fix the issue and upgrade to 4.5-RC1 successfully?

@wolfs 

**Command**
./gradlew -S :proto-converter:generateProto

**Result**

```
The :proto-converter:generateProto task failed.
Multiple input file properties with name '$1'
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':proto-converter:generateProto'.Open stacktrace
Caused by: java.lang.IllegalArgumentException: Multiple input file properties with name '$1'Close stacktrace
at org.gradle.api.internal.tasks.properties.GetInputFilesVisitor.visitInputFileProperty(GetInputFilesVisitor.java:44)
at org.gradle.api.internal.tasks.properties.CompositePropertyVisitor.visitInputFileProperty(CompositePropertyVisitor.java:35)
at org.gradle.api.internal.tasks.DefaultTaskInputs.visitRegisteredProperties(DefaultTaskInputs.java:76)
at org.gradle.api.internal.tasks.TaskPropertyUtils.visitProperties(TaskPropertyUtils.java:39)
at org.gradle.api.internal.tasks.execution.DefaultTaskProperties.resolve(DefaultTaskProperties.java:77)
at org.gradle.execution.taskgraph.TaskInfo.getTaskProperties(TaskInfo.java:229)
at org.gradle.execution.taskgraph.TaskInfo.getDestroyables(TaskInfo.java:220)
at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.getDestroyablePaths(DefaultTaskExecutionPlan.java:794)
at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.canRunWithCurrentlyExecutedTasks(DefaultTaskExecutionPlan.java:663)
at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.access$2600(DefaultTaskExecutionPlan.java:89)
at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan$9.transform(DefaultTaskExecutionPlan.java:593)
at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan$9.transform(DefaultTaskExecutionPlan.java:588)
at org.gradle.internal.resources.DefaultResourceLockCoordinationService.withStateLock(DefaultResourceLockCoordinationService.java:45)
at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.selectNextTask(DefaultTaskExecutionPlan.java:588)
at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.access$2300(DefaultTaskExecutionPlan.java:89)
at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan$8.transform(DefaultTaskExecutionPlan.java:562)
at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan$8.transform(DefaultTaskExecutionPlan.java:543)
at org.gradle.internal.resources.DefaultResourceLockCoordinationService.withStateLock(DefaultResourceLockCoordinationService.java:45)
at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:543)
at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:98)
at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
at java.lang.Thread.run(Thread.java:748)
```

--- a/subprojects/core/src/integTest/groovy/org/gradle/api/internal/project/taskfactory/TaskPropertyNamingIntegrationTest.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/internal/project/taskfactory/TaskPropertyNamingIntegrationTest.groovy
@@ -245,6 +245,77 @@ class TaskPropertyNamingIntegrationTest extends AbstractIntegrationSpec {
         output =~ /Local state: '.*localState'/
     }
 
+    def "unnamed file properties are named"() {
+        buildFile << """
+            import org.gradle.api.internal.tasks.*
+            import org.gradle.api.internal.tasks.properties.*
+
+            task myTask {
+                inputs.file("input.txt")
+                inputs.files("input-1.txt", "input-2.txt")
+                inputs.dir("input-dir")
+
+                outputs.file("output.txt")
+                outputs.files("output-1.txt", "output-2.txt")
+                outputs.dir("output-dir")
+                outputs.dirs("output-dir-1", "output-dir-2")
+            }
+        """ << printProperties("myTask")
+        when:
+        run "printProperties"
+        then:
+        output.contains """
+            Input file property '\$1'
+            Input file property '\$2'
+            Input file property '\$3'
+            Output file property '\$1'
+            Output file property '\$2'
+            Output file property '\$3'
+            Output file property '\$4'
+        """.stripIndent()
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/4085")
+    def "can register more unnamed properties after properties have been queried"() {
+        buildFile << """
+            import org.gradle.api.internal.tasks.*
+            import org.gradle.api.internal.tasks.properties.*
+
+            task myTask {
+                // Register first unnamed property
+                inputs.file("input-1.txt")
+                // Trigger calculating unnamed property names
+                inputs.hasInputs
+
+                // Register another unnamed property
+                inputs.file("input-2.txt")
+                // Trigger calculating unnamed property names again
+                inputs.hasInputs
+                
+                // Register first unnamed property
+                outputs.file("output-1.txt")
+                // Trigger calculating unnamed property names
+                outputs.hasOutput
+
+                // Register another unnamed property
+                outputs.file("output-2.txt")
+                // Trigger calculating unnamed property names again
+                outputs.hasOutput
+            }
+            
+            ${printProperties("myTask")}
+        """
+        when:
+        run "printProperties"
+        then:
+        output.contains """
+            Input file property '\$1'
+            Input file property '\$2'
+            Output file property '\$1'
+            Output file property '\$2'
+        """.stripIndent()
+    }
+
     def "input properties can be overridden"() {
         buildFile << classesForNestedProperties()
         buildFile << """
@@ -347,6 +418,14 @@ class TaskPropertyNamingIntegrationTest extends AbstractIntegrationSpec {
                         }
                     })
                 }
+            }
+        """
+    }
+
+    def printProperties(String task) {
+        printPropertiesTask() << """
+            task printProperties(type: PrintInputsAndOutputs) {
+                task = $task
             }
         """
     }

--- a/subprojects/core/src/main/java/org/gradle/api/internal/FilePropertyContainer.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/FilePropertyContainer.java
@@ -1,0 +1,61 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal;
+
+import com.google.common.collect.Lists;
+import org.gradle.api.internal.tasks.TaskFilePropertySpec;
+import org.gradle.api.tasks.TaskFilePropertyBuilder;
+
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Container for {@link org.gradle.api.internal.tasks.TaskPropertySpec}s that might not have a name. The container
+ * ensures that whenever properties are iterated they are always assigned a name.
+ */
+public class FilePropertyContainer<T extends TaskFilePropertyBuilder & TaskFilePropertySpec> implements Iterable<T> {
+    private final List<T> properties = Lists.newArrayList();
+    private boolean changed;
+    private int unnamedPropertyCounter;
+
+    private FilePropertyContainer() {
+    }
+
+    public static <T extends TaskFilePropertyBuilder & TaskFilePropertySpec> FilePropertyContainer<T> create() {
+        return new FilePropertyContainer<T>();
+    }
+
+    public void add(T property) {
+        properties.add(property);
+        changed = true;
+    }
+
+    @Override
+    public Iterator<T> iterator() {
+        if (changed) {
+            for (T propertySpec : properties) {
+                String propertyName = propertySpec.getPropertyName();
+                if (propertyName == null) {
+                    propertyName = "$" + (++unnamedPropertyCounter);
+                    propertySpec.withPropertyName(propertyName);
+                }
+            }
+            changed = false;
+        }
+        return properties.iterator();
+    }
+}

--- a/subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskInputs.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskInputs.java
@@ -19,6 +19,7 @@ import com.google.common.collect.Lists;
 import org.gradle.api.Describable;
 import org.gradle.api.NonNullApi;
 import org.gradle.api.file.FileCollection;
+import org.gradle.api.internal.FilePropertyContainer;
 import org.gradle.api.internal.TaskInputsInternal;
 import org.gradle.api.internal.TaskInternal;
 import org.gradle.api.internal.file.CompositeFileCollection;
@@ -37,8 +38,6 @@ import java.util.List;
 import java.util.Map;
 import java.util.concurrent.Callable;
 
-import static org.gradle.api.internal.tasks.TaskPropertyUtils.ensurePropertiesHaveNames;
-
 @NonNullApi
 public class DefaultTaskInputs implements TaskInputsInternal {
     private final FileCollection allInputFiles;
@@ -47,7 +46,7 @@ public class DefaultTaskInputs implements TaskInputsInternal {
     private final TaskMutator taskMutator;
     private final PropertyWalker propertyWalker;
     private final List<DeclaredTaskInputProperty> registeredProperties = Lists.newArrayList();
-    private final List<DeclaredTaskInputFileProperty> registeredFileProperties = Lists.newArrayList();
+    private final FilePropertyContainer<DeclaredTaskInputFileProperty> registeredFileProperties = FilePropertyContainer.create();
     private final TaskInputs deprecatedThis;
     private final PropertySpecFactory specFactory;
 
@@ -71,7 +70,6 @@ public class DefaultTaskInputs implements TaskInputsInternal {
 
     @Override
     public void visitRegisteredProperties(PropertyVisitor visitor) {
-        ensurePropertiesHaveNames(registeredFileProperties);
         for (DeclaredTaskInputFileProperty fileProperty : registeredFileProperties) {
             visitor.visitInputFileProperty(fileProperty);
         }
@@ -133,7 +131,7 @@ public class DefaultTaskInputs implements TaskInputsInternal {
 
     @Override
     public boolean getHasSourceFiles() {
-        GetInputFilesVisitor visitor = new GetInputFilesVisitor(task.toString());
+        GetInputFilesVisitor visitor = new GetInputFilesVisitor();
         TaskPropertyUtils.visitProperties(propertyWalker, task, visitor);
         return visitor.hasSourceFiles();
     }

--- a/subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskOutputs.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskOutputs.java
@@ -17,12 +17,12 @@
 package org.gradle.api.internal.tasks;
 
 import com.google.common.collect.ImmutableSortedSet;
-import com.google.common.collect.Lists;
 import groovy.lang.Closure;
 import org.gradle.api.Describable;
 import org.gradle.api.NonNullApi;
 import org.gradle.api.Task;
 import org.gradle.api.file.FileCollection;
+import org.gradle.api.internal.FilePropertyContainer;
 import org.gradle.api.internal.OverlappingOutputs;
 import org.gradle.api.internal.TaskExecutionHistory;
 import org.gradle.api.internal.TaskInternal;
@@ -62,7 +62,7 @@ public class DefaultTaskOutputs implements TaskOutputsInternal {
     private List<SelfDescribingSpec<TaskInternal>> cacheIfSpecs = new LinkedList<SelfDescribingSpec<TaskInternal>>();
     private List<SelfDescribingSpec<TaskInternal>> doNotCacheIfSpecs = new LinkedList<SelfDescribingSpec<TaskInternal>>();
     private TaskExecutionHistory history;
-    private final List<DeclaredTaskOutputFileProperty> registeredFileProperties = Lists.newArrayList();
+    private final FilePropertyContainer<DeclaredTaskOutputFileProperty> registeredFileProperties = FilePropertyContainer.create();
     private final TaskInternal task;
     private final TaskMutator taskMutator;
 
@@ -76,7 +76,6 @@ public class DefaultTaskOutputs implements TaskOutputsInternal {
 
     @Override
     public void visitRegisteredProperties(PropertyVisitor visitor) {
-        TaskPropertyUtils.ensurePropertiesHaveNames(registeredFileProperties);
         for (DeclaredTaskOutputFileProperty fileProperty : registeredFileProperties) {
             visitor.visitOutputFileProperty(fileProperty);
         }

--- a/subprojects/core/src/main/java/org/gradle/api/internal/tasks/TaskPropertyUtils.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/tasks/TaskPropertyUtils.java
@@ -23,7 +23,6 @@ import org.gradle.api.internal.TaskInternal;
 import org.gradle.api.internal.project.ProjectInternal;
 import org.gradle.api.internal.tasks.properties.PropertyVisitor;
 import org.gradle.api.internal.tasks.properties.PropertyWalker;
-import org.gradle.api.tasks.TaskFilePropertyBuilder;
 import org.gradle.internal.Cast;
 
 import java.util.Iterator;
@@ -61,17 +60,6 @@ public class TaskPropertyUtils {
             builder.add(propertySpec);
         }
         return builder.build();
-    }
-
-    public static <T extends TaskPropertySpec & TaskFilePropertyBuilder> void ensurePropertiesHaveNames(Iterable<T> properties) {
-        int unnamedPropertyCounter = 0;
-        for (T propertySpec : properties) {
-            String propertyName = propertySpec.getPropertyName();
-            if (propertyName == null) {
-                propertyName = "$" + (++unnamedPropertyCounter);
-                propertySpec.withPropertyName(propertyName);
-            }
-        }
     }
 
     public static <T extends TaskFilePropertySpec> SortedSet<ResolvedTaskOutputFilePropertySpec> resolveFileProperties(ImmutableSortedSet<T> properties) {

--- a/subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/DefaultTaskProperties.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/DefaultTaskProperties.java
@@ -67,7 +67,7 @@ public class DefaultTaskProperties implements TaskProperties {
 
     public static TaskProperties resolve(PropertyWalker propertyWalker, PathToFileResolver resolver, TaskInternal task) {
         String beanName = task.toString();
-        GetInputFilesVisitor inputFilesVisitor = new GetInputFilesVisitor(beanName);
+        GetInputFilesVisitor inputFilesVisitor = new GetInputFilesVisitor();
         GetOutputFilesVisitor outputFilesVisitor = new GetOutputFilesVisitor();
         GetInputPropertiesVisitor inputPropertiesVisitor = new GetInputPropertiesVisitor(beanName);
         GetLocalStateVisitor localStateVisitor = new GetLocalStateVisitor(beanName, resolver);

--- a/subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/GetInputFilesVisitor.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/GetInputFilesVisitor.java
@@ -17,33 +17,24 @@
 package org.gradle.api.internal.tasks.properties;
 
 import com.google.common.collect.ImmutableSortedSet;
-import com.google.common.collect.Sets;
-import org.gradle.api.file.FileCollection;
-import org.gradle.api.internal.file.CompositeFileCollection;
-import org.gradle.api.internal.file.collections.FileCollectionResolveContext;
+import com.google.common.collect.Lists;
 import org.gradle.api.internal.tasks.TaskInputFilePropertySpec;
+import org.gradle.api.internal.tasks.TaskPropertyUtils;
 
-import java.util.Set;
+import java.util.List;
 
 public class GetInputFilesVisitor extends PropertyVisitor.Adapter {
-    private final String beanName;
-    private ImmutableSortedSet.Builder<TaskInputFilePropertySpec> builder = ImmutableSortedSet.naturalOrder();
-    private Set<String> names = Sets.newHashSet();
+    private final List<TaskInputFilePropertySpec> specs = Lists.newArrayList();
     private boolean hasSourceFiles;
 
     private ImmutableSortedSet<TaskInputFilePropertySpec> fileProperties;
 
-    public GetInputFilesVisitor(String beanName) {
-        this.beanName = beanName;
+    public GetInputFilesVisitor() {
     }
 
     @Override
     public void visitInputFileProperty(TaskInputFilePropertySpec inputFileProperty) {
-        String propertyName = inputFileProperty.getPropertyName();
-        if (!names.add(propertyName)) {
-            throw new IllegalArgumentException(String.format("Multiple %s file properties with name '%s'", "input", propertyName));
-        }
-        builder.add(inputFileProperty);
+        specs.add(inputFileProperty);
         if (inputFileProperty.isSkipWhenEmpty()) {
             hasSourceFiles = true;
         }
@@ -51,43 +42,9 @@ public class GetInputFilesVisitor extends PropertyVisitor.Adapter {
 
     public ImmutableSortedSet<TaskInputFilePropertySpec> getFileProperties() {
         if (fileProperties == null) {
-            fileProperties = builder.build();
+            fileProperties = TaskPropertyUtils.collectFileProperties("input", specs.iterator());
         }
         return fileProperties;
-    }
-
-    public FileCollection getFiles() {
-        return new CompositeFileCollection() {
-            @Override
-            public String getDisplayName() {
-                return beanName + " input files";
-            }
-
-            @Override
-            public void visitContents(FileCollectionResolveContext context) {
-                for (TaskInputFilePropertySpec filePropertySpec : getFileProperties()) {
-                    context.add(filePropertySpec.getPropertyFiles());
-                }
-            }
-        };
-    }
-
-    public FileCollection getSourceFiles() {
-        return new CompositeFileCollection() {
-            @Override
-            public String getDisplayName() {
-                return beanName + " source files";
-            }
-
-            @Override
-            public void visitContents(FileCollectionResolveContext context) {
-                for (TaskInputFilePropertySpec filePropertySpec : getFileProperties()) {
-                    if (filePropertySpec.isSkipWhenEmpty()) {
-                        context.add(filePropertySpec.getPropertyFiles());
-                    }
-                }
-            }
-        };
     }
 
     public boolean hasSourceFiles() {

--- a/subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/GetOutputFilesVisitor.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/GetOutputFilesVisitor.java
@@ -18,6 +18,7 @@ package org.gradle.api.internal.tasks.properties;
 
 import com.google.common.collect.ImmutableSortedSet;
 import com.google.common.collect.Iterators;
+import com.google.common.collect.Lists;
 import org.gradle.api.NonNullApi;
 import org.gradle.api.internal.tasks.CacheableTaskOutputFilePropertySpec;
 import org.gradle.api.internal.tasks.CompositeTaskOutputPropertySpec;
@@ -25,12 +26,11 @@ import org.gradle.api.internal.tasks.TaskOutputFilePropertySpec;
 import org.gradle.api.internal.tasks.TaskPropertyUtils;
 
 import java.io.File;
-import java.util.ArrayList;
 import java.util.List;
 
 @NonNullApi
 public class GetOutputFilesVisitor extends PropertyVisitor.Adapter {
-    private List<TaskOutputFilePropertySpec> specs = new ArrayList<TaskOutputFilePropertySpec>();
+    private final List<TaskOutputFilePropertySpec> specs = Lists.newArrayList();
     private ImmutableSortedSet<TaskOutputFilePropertySpec> fileProperties;
     private boolean hasDeclaredOutputs;
 

--- a/subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/DefaultTaskInputsTest.groovy
+++ b/subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/DefaultTaskInputsTest.groovy
@@ -25,8 +25,11 @@ import org.gradle.api.internal.tasks.properties.DefaultPropertyMetadataStore
 import org.gradle.api.internal.tasks.properties.DefaultPropertyWalker
 import org.gradle.api.internal.tasks.properties.GetInputFilesVisitor
 import org.gradle.api.internal.tasks.properties.GetInputPropertiesVisitor
+import org.gradle.api.internal.tasks.properties.PropertyVisitor
 import org.gradle.util.UsesNativeServices
+import spock.lang.Issue
 import spock.lang.Specification
+import spock.lang.Unroll
 
 import java.util.concurrent.Callable
 
@@ -282,6 +285,29 @@ class DefaultTaskInputsTest extends Specification {
         inputs.hasSourceFiles
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/4085")
+    @Unroll
+    def "can register more unnamed properties with method #method after properties have been queried"() {
+        inputs."$method"("input-1")
+        // Trigger naming properties
+        inputs.hasSourceFiles
+        inputs."$method"("input-2")
+        def names = []
+
+        when:
+        inputs.visitRegisteredProperties(new PropertyVisitor.Adapter() {
+            @Override
+            void visitInputFileProperty(TaskInputFilePropertySpec property) {
+                names += property.propertyName
+            }
+        })
+        then:
+        names == ['$1', '$2']
+
+        where:
+        method << ["file", "dir", "files"]
+    }
+
     def inputProperties() {
         def visitor = new GetInputPropertiesVisitor("test")
         TaskPropertyUtils.visitProperties(walker, task, visitor)
@@ -289,7 +315,7 @@ class DefaultTaskInputsTest extends Specification {
     }
 
     def inputFileProperties() {
-        GetInputFilesVisitor visitor = new GetInputFilesVisitor("test")
+        GetInputFilesVisitor visitor = new GetInputFilesVisitor()
         TaskPropertyUtils.visitProperties(walker, task, visitor)
         return visitor.getFileProperties()
     }

--- a/subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/DefaultTaskOutputsTest.groovy
+++ b/subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/DefaultTaskOutputsTest.groovy
@@ -27,7 +27,9 @@ import org.gradle.api.internal.project.ProjectInternal
 import org.gradle.api.internal.tasks.execution.TaskProperties
 import org.gradle.api.internal.tasks.properties.DefaultPropertyMetadataStore
 import org.gradle.api.internal.tasks.properties.DefaultPropertyWalker
+import org.gradle.api.internal.tasks.properties.PropertyVisitor
 import org.gradle.util.UsesNativeServices
+import spock.lang.Issue
 import spock.lang.Specification
 import spock.lang.Unroll
 
@@ -251,6 +253,29 @@ class DefaultTaskOutputsTest extends Specification {
         then:
         GradleException e = thrown()
         e.message.contains("Could not evaluate spec for 'Exception is thrown'.")
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/4085")
+    @Unroll
+    def "can register more unnamed properties with method #method after properties have been queried"() {
+        outputs."$method"("output-1")
+        // Trigger naming properties
+        outputs.hasOutput
+        outputs."$method"("output-2")
+        def names = []
+
+        when:
+        outputs.visitRegisteredProperties(new PropertyVisitor.Adapter() {
+            @Override
+            void visitOutputFileProperty(TaskOutputFilePropertySpec property) {
+                names += property.propertyName
+            }
+        })
+        then:
+        names == ['$1', '$2']
+
+        where:
+        method << ["file", "dir", "files", "dirs"]
     }
 
     void canRegisterOutputFiles() {

